### ISSUE TITLE
BSOD when using a client instantiated with "localhost" hostname

### ISSUE THREAD

**Issue by gsferreira** [grpc/grpc#18057]
I have a use case where an application starts another process that is a gRPC server to which the first application is a client.
Recently I saw that sometimes, in the first request from the Client to the Server, the test machine has a blue screen (BSOD).

I've been trying to figure out what I was doing wrong and after a few tests I finally figure out that this is only happening if the client tries to call the server using the URI "localhost:YYYY" instead of "127.0.0.1:YYYY".

I've been trying to find some kind of reason in gRPC documentation to have the 127.0.0.1 in the client, but I can't find anything. 

### What version of gRPC and what language are you using?

1.18.0
 
### What operating system (Linux, Windows, …) and version?

Microsoft Windows 10 Pro Version	10.0.17134
 
### What did you do?

1. Download the C# Greeter example
1. Upgrade to 1.18.0
1. Change _GreeterClient.csproj_ and _GreeterServer.csproj_ TargetFramework to net472
1. Update the GreeterServer program.cs with
```c#
class GreeterImpl : Greeter.GreeterBase
    {
        private readonly TaskCompletionSource<object> _shutdownTcs = new TaskCompletionSource<object>();
        public Task ShutdownTask => _shutdownTcs.Task;

        // Server side handler of the SayHello RPC
        public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
        {
            _shutdownTcs.SetResult(null);
            return Task.FromResult(new HelloReply { Message = "Hello " + request.Name });
        }
    }

    class Program
    {
        const int Port = 65051;

        public static void Main(string[] args)
        {
            var service = new GreeterImpl();

            Server server = new Server
            {
                Services = { Greeter.BindService(service) },
                Ports = { new ServerPort("localhost", Port, ServerCredentials.Insecure) }
            };
            server.Start();

            Console.WriteLine("Greeter server listening on port " + Port);

            service.ShutdownTask.Wait();

            server.ShutdownAsync().Wait();
        }
    }
```
5. Upgrade GreeterClient program.cs with

```c#
   class Program
    {
        public static void Main(string[] args)
        {
            var serverId = StartServer().GetAwaiter().GetResult();

            Channel channel = new Channel("localhost:65051", ChannelCredentials.Insecure);

            var client = new Greeter.GreeterClient(channel);
            String user = "Luke";

            var reply = client.SayHello(new HelloRequest { Name = user });
            Console.WriteLine($"Greeting from {serverId}: {reply.Message}");

            channel.ShutdownAsync().Wait();

        }

        private static Task<int> StartServer()
        {
            var tcs = new TaskCompletionSource<int>();

            var serverProcess = CreateServer();

            
            serverProcess.OutputDataReceived += (s, e) =>
            {
                if (e.Data == null || !e.Data.Contains("Greeter server listening on port")) return;

                tcs.SetResult(serverProcess.Id);    
            };

            serverProcess.Start();
            serverProcess.BeginOutputReadLine();

            if (serverProcess.HasExited)
            {
                tcs.SetException(
                    new Exception("Can't start server"));
            }

            return tcs.Task;
        }

        private static Process CreateServer()
        {
            var path =
                @"..\..\..\..\GreeterServer\bin\Release\net472\GreeterServer.exe";

            var serverProcess = new Process
            {
                StartInfo = new ProcessStartInfo(path)
                {
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardOutput = true,
                }
            };

            return serverProcess;
        }
```

6. Compile to Release
7. Run the GreeterClient with the following .bat file

```
:loop
start "Client" GreeterClient.exe
timeout /t 3
goto loop
```
8. Run it and wait
 
### What did you expect to see?

I was expecting to connect with no trouble in every attempt.
At least, the server shouldn't shutdown.
 
### What did you see instead?

The server shutdowns with a blue screen.
In the event viewer I have a message from Windows Error Reporting saying:
_Fault bucket AV_tcpip!TcpDeliverLoopbackDataToClient, type 0_

 
### Anything else we should know about your project / environment?

Virtual Machine running on Azure.
Size: Basic A2 (2 vcpus, 3.5 GB memory)
Processor:	Intel(R) Xeon(R) CPU E5-2660 0 @ 2.20GHz, 2200 Mhz, 2 Core(s), 2 Logical Processor(s)

**Comment by nycdotnet** [comment]
We've been dealing with this BSOD issue at Namely for almost a year and never expected it might be this.  I'm excited to hear that there may be a workaround by using 127.0.0.1 instead of localhost.

**Comment by ricardoschroeder** [comment]
We've also been facing this issue for about a month or so now. I've been running some tests and it seems to be working. I still need to run more performance + multi-thread tests. 
Thanks @gsferreira

**Comment by jtattermusch** [comment]
To be honest, I have trouble understanding how gRPC can be the cause of a server getting BSOD - there really seems like a problem of the operating system or the VM, not gRPC. If this is a grpc problem, shouldn't the process just crash (ideally with a coredump that can be used to diagnose the problem)?
Also, can you reproduce this on a physical Windows machine or is it just Azure VM?

**Comment by nycdotnet** [comment]
We get this on our Windows 10 x64 workstations. I’m sure it’s not “gRPC” but clearly something about what it’s doing is causing a problem in the networking stack and possibly more narrowly the DNS client.  Since it’s clearly not just us, I’ll see if I can get it to happen with the suggested repro and if so see if we can get a dump over to MS to get this solved.

**Comment by jtattermusch** [comment]
Having a repro that works on Win 10 workstation would be great.

**Comment by nicolasnoble** [comment]
Alternatively, do you have any information FROM the bsod? It usually has some details about what crashed where.

**Comment by nicolasnoble** [comment]
Also, yes, a BSOD clearly is on Microsoft's, even if this is somehow linked to some usage pattern that grpc does on their network stack.

**Comment by nycdotnet** [comment]
I was able to get my Windows 10 x64 workstation to BSOD once last night this using the grpc greeter example project modified to use the example code from @gsferreira (with a few async/path fixes).  If you're interested, the code is implemented in the latest two commits to the `repro-bsod` branch of my fork of grpc.

https://github.com/nycdotnet/grpc/tree/repro-bsod

The bluescreen referenced tcpip.sys and the stop code was `DRIVER_IRQL_NOT_LESS_OR_EQUAL`.

It doesn't happen all the time, and when I had gotten it to repro, I had just rebooted.  I have the minidump and I'm going to forward to my contacts at Microsoft.  I don't know if there's much that can be done by the gRPC project for this, but will report back if I hear anything.

**Comment by nicolasnoble** [comment]
This bsod is a driver error. Since this touches localhost, this might be a bug in the actual tcp stack of windows. This might be linked to the way we're querying inconditionally the kernel for the driver of the interface we're told to use.

https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/tcp_client_windows.cc#L162

**Comment by jtattermusch** [comment]
Seems like I'm able to reproduce by just running grpc tests (e.g csharp or python) on a kokoro win2016 VM. It doesn't happen 100%, but often enough.

**Comment by jtattermusch** [comment]
Actually, the problem I'm seeing might be something else - it looks like I'm getting a BSOD when building :-( I'll update one I figure out what's going on.

**Comment by jtattermusch** [comment]
I also reached out to my Microsoft contacts, I'll update if there's any news.

**Comment by nicolasnoble** [comment]
This is an out of bound memset in the function `IppGetNextAddress` of tcpip.sys, right after constructing a new IRQ.

![image](https://user-images.githubusercontent.com/7281574/53598173-e6643b00-3b58-11e9-84b1-2f7e52607e72.png)

**Comment by nicolasnoble** [comment]
This might be quite an old bug from Microsoft:

https://www.bleepingcomputer.com/forums/t/543317/bsod-driver-irql-not-less-or-equal-dump-file-attached/

**Comment by nicolasnoble** [comment]
Downgrading the priority on this one since there's quite nothing we can do on our side but waiting on Microsoft for a fix in their kernel.

**Comment by jtattermusch** [comment]
Very interesting findings:
I ran a few experiments with kokoro Win2016 machines and I seem to have pinned the problem down to https://github.com/grpc/grpc/pull/14905 (I just guessed, sometimes you need to follow your instincts ;-) )
* Without that change, I can run the entire C-core test suite without problems on a win2016 machine and it ends up green.
* With that change, I seem to be getting a BSOD (a minidump from a remote machine) pretty reliably.

**Comment by jtattermusch** [comment]
@nycdotnet  @gsferreira   would you be able to tests if it's really #14905 that's causing trouble on your side? Depending on the language you use, we can provide you with an adhoc package build so you have something to test against.

**Comment by nycdotnet** [comment]
We use c#.  I would be willing to test this.

**Comment by nicolasnoble** [comment]
And the dilemma here is that #14905 is drastically improving local communication performance.

**Comment by jtattermusch** [comment]
@nycdotnet  can you please test using the nuget packages from the following comment and post back the results? https://github.com/grpc/grpc/pull/18744#issuecomment-482635645

**Comment by yulia-d** [comment]
@jtattermusch We are having BSODs during local communication on Windows Server 2012 R2 Standard (64 bit) **and the package you posted resolves them**.

**Comment by nycdotnet** [comment]
FYI: I am working on this, but it may take a bit as I yak shave my way through CI and private NuGet stuff.

**Comment by nycdotnet** [comment]
@jtattermusch I concur that this appears to have fixed it - I've run my test suite on Windows against `localhost` three times now with my helpers library compiled against the below and have had no BSOD.  It was rare to be able to get through it once before.

- Google.Protobuf  3.7.0
- Grpc 1.21.0-dev201904121506
- Grpc.HealthCheck 1.21.0-dev201904121506
- Grpc.Reflection 1.21.0-dev201904121506
- System.Runtime.Loader 4.3.0 (due to this having been removed as a dependency from one of the above packages)

We do not reference the other packages.

**Comment by yulia-d** [comment]
BTW I got rid of all mentions of "localhost" in our code and replaced it with loop-back and it still crashes, so the work-around mentioned above doesn't really help.

**Comment by mrutter** [comment]
Just to let you know that I have the same issue here: BSOD on Windows Server 2016 Datacenter x64. The server seems to crash after a period of usage. Since we installed the service based on gRPC, the server crashed three times. I'm collecting more information and I will write back asap.

Thank you.

**Comment by vjpai** [comment]
@nicolasnoble in response to your comment: #14905 only improves the performance on the loopback interface, correct? Although that's nice for microbenchmarks, it's not necessarily a realistic operating point so it wouldn't really break the world to roll it back if it's indeed the source of instability (IMO), amirite?

**Comment by hacst** [comment]
I can verify that this is an actual bug in Windows fast tcp loopback. I reported this to the Microsoft Security Response Center as a potential security issue at the start of the month after investigating BSODs on Windows Server 2016 machines. Sadly I did not see this issue at the time. They were able to reproduce the issue based on a small sample I created and verified that it is caused by fast loopback.

As it is a local DOS they judged the issue to be of moderate security impact which they wouldn't have to address in a security bulletin. Their recommendation is to stop using SIO_LOOPBACK_FAST_PATH. It is deprecated by now anyways (see https://docs.microsoft.com/en-us/windows-hardware/drivers/network/sio-loopback-fast-path ) so removing it is exactly the right thing to do.

MSRC Case number was 51262
